Switching between pnpm and npm is broken on first try when cleaning the project

To reproduce:
1) take a v20 Vaadin Spring Boot project, test that it works
2) swap to npm by adding `vaadin.pnpm.enable =false` to `application.properties`
3) clean the project with `mvn clean`
4) rerun the project.
Steps 3+4 can be combined with `mvn clean spring-boot:run`.

Expected: project works

Actual: npm/pnpm install fails due to missing `target/flow-frontend/package.json` 

Rerunning the project will work.

It seems that there is some broken logic in
1) https://github.com/vaadin/flow/blob/master/vaadin-dev-server/src/main/java/com/vaadin/base/devserver/startup/DevModeInitializer.java#L351
2) https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGeneratePackageJson.java#L68

but neither of these are a problem because they just make the `TaskGeneratePackgeJson` always create the file, but in the end something is still removing it, - investigation to be continued.

This is still broken in commons-io 2.11.0 and Vaadin 22.0.0.rc1

```
[qtp64797628-30] ERROR com.vaadin.flow.server.DefaultErrorHandler -
java.lang.IllegalStateException: com.vaadin.flow.server.ExecutionFailedException: Couldn't remove C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules directory
        at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:772)
        at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:768)
        at com.vaadin.base.devserver.AbstractDevServerRunner.handleRequest(AbstractDevServerRunner.java:634)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1567)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:299)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:228)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1435)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1350)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.vaadin.flow.server.ExecutionFailedException: Couldn't remove C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules directory
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.runNpmInstall(TaskRunNpmInstall.java:315)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.execute(TaskRunNpmInstall.java:153)
        at com.vaadin.flow.server.frontend.NodeTasks.execute(NodeTasks.java:913)
        at com.vaadin.base.devserver.startup.DevModeInitializer.runNodeTasks(DevModeInitializer.java:397)
        at com.vaadin.base.devserver.startup.DevModeInitializer.lambda$initDevModeHandler$0(DevModeInitializer.java:326)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1728)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.io.IOException: Cannot delete file: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules
        at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1344)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.cleanUp(TaskRunNpmInstall.java:548)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.runNpmInstall(TaskRunNpmInstall.java:313)
        ... 11 more
Caused by: java.nio.file.NoSuchFileException: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules\.pnpm\@babel\core@7.16.0\node_modules\@babel\code-frame
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsDirectoryStream.<init>(WindowsDirectoryStream.java:86)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.newDirectoryStream(WindowsFileSystemProvider.java:523)
        at java.base/java.nio.file.Files.newDirectoryStream(Files.java:472)
        at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:300)
        at java.base/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:373)
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2761)
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2797)
        at org.apache.commons.io.file.PathUtils.visitFileTree(PathUtils.java:971)
        at org.apache.commons.io.file.PathUtils.deleteDirectory(PathUtils.java:434)
        at org.apache.commons.io.file.PathUtils.delete(PathUtils.java:391)
        at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1341)
        ... 13 more
[WARNING] handleException / java.io.IOException: Cannot delete file: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules
[qtp64797628-28] ERROR com.vaadin.flow.server.DefaultErrorHandler -
java.lang.IllegalStateException: com.vaadin.flow.server.ExecutionFailedException: Couldn't remove C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules directory
        at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:772)
        at com.vaadin.base.devserver.AbstractDevServerRunner.getCause(AbstractDevServerRunner.java:768)
        at com.vaadin.base.devserver.AbstractDevServerRunner.handleRequest(AbstractDevServerRunner.java:634)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1567)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:299)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
        at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1626)
        at org.eclipse.jetty.websocket.server.WebSocketUpgradeFilter.doFilter(WebSocketUpgradeFilter.java:228)
        at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
        at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1435)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1350)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:516)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
        at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.vaadin.flow.server.ExecutionFailedException: Couldn't remove C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules directory
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.runNpmInstall(TaskRunNpmInstall.java:315)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.execute(TaskRunNpmInstall.java:153)
        at com.vaadin.flow.server.frontend.NodeTasks.execute(NodeTasks.java:913)
        at com.vaadin.base.devserver.startup.DevModeInitializer.runNodeTasks(DevModeInitializer.java:397)
        at com.vaadin.base.devserver.startup.DevModeInitializer.lambda$initDevModeHandler$0(DevModeInitializer.java:326)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1728)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.io.IOException: Cannot delete file: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules
        at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1344)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.cleanUp(TaskRunNpmInstall.java:548)
        at com.vaadin.flow.server.frontend.TaskRunNpmInstall.runNpmInstall(TaskRunNpmInstall.java:313)
        ... 11 more
Caused by: java.nio.file.NoSuchFileException: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules\.pnpm\@babel\core@7.16.0\node_modules\@babel\code-frame
        at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:85)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
        at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
        at java.base/sun.nio.fs.WindowsDirectoryStream.<init>(WindowsDirectoryStream.java:86)
        at java.base/sun.nio.fs.WindowsFileSystemProvider.newDirectoryStream(WindowsFileSystemProvider.java:523)
        at java.base/java.nio.file.Files.newDirectoryStream(Files.java:472)
        at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:300)
        at java.base/java.nio.file.FileTreeWalker.next(FileTreeWalker.java:373)
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2761)
        at java.base/java.nio.file.Files.walkFileTree(Files.java:2797)
        at org.apache.commons.io.file.PathUtils.visitFileTree(PathUtils.java:971)
        at org.apache.commons.io.file.PathUtils.deleteDirectory(PathUtils.java:434)
        at org.apache.commons.io.file.PathUtils.delete(PathUtils.java:391)
        at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1341)
        ... 13 more
[WARNING] handleException /sw.js java.io.IOException: Cannot delete file: C:\Users\Pekka\workspaces\flow\skeleton-starter-flow\node_modules
```

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -18,6 +18,7 @@ package com.vaadin.flow.server.frontend;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.Serializable;
@@ -26,18 +27,18 @@ import java.net.HttpURLConnection;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
+import java.nio.file.LinkOption;
 import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
 import java.time.LocalDateTime;
 import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Scanner;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import org.apache.commons.io.FileUtils;
@@ -1259,15 +1260,78 @@ public class FrontendUtils {
                     + " does not look like a node_modules directory");
         }
 
-        Path nodeModulesPath = nodeModules.toPath();
-        try (Stream<Path> walk = Files.walk(nodeModulesPath)) {
-            String undeletable = walk.sorted(Comparator.reverseOrder())
-                    .map(Path::toFile).filter(file -> !file.delete())
-                    .map(File::getAbsolutePath)
-                    .collect(Collectors.joining(", "));
+        deleteDirectory(nodeModules);
+    }
 
-            if (!undeletable.isEmpty() && nodeModules.exists()) {
-                throw new IOException("Unable to delete files: " + undeletable);
+    public static void deleteDirectory(File directory) throws IOException {
+        if (!directory.exists() || !directory.isDirectory()) {
+            return;
+        }
+
+        if (!(Files.isSymbolicLink(directory.toPath()) || isJunction(directory.toPath()))) {
+            cleanDirectory(directory);
+        }
+
+        if (!directory.delete()) {
+            String message = "Unable to delete directory " + directory + ".";
+            throw new IOException(message);
+        }
+    }
+
+    /**
+     * Check that directory is not a windows junction which is basically a symlink.
+     * @param directory directory path to check
+     * @return true if directory is a windows junction
+     * @throws IOException if an I/O error occurs
+     */
+    private static boolean isJunction(Path directory) throws IOException {
+        boolean isWindows = System.getProperty("os.name").toLowerCase().contains("windows");
+        BasicFileAttributes attrs = Files.readAttributes(directory, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
+        return isWindows && attrs.isDirectory() && attrs.isOther();
+    }
+
+    private static void cleanDirectory(File directory) throws IOException {
+        if (!directory.exists()) {
+            String message = directory + " does not exist";
+            throw new IllegalArgumentException(message);
+        }
+
+        if (!directory.isDirectory()) {
+            String message = directory + " is not a directory";
+            throw new IllegalArgumentException(message);
+        }
+
+        File[] files = directory.listFiles();
+        if (files == null) {  // null if security restricted
+            throw new IOException("Failed to list contents of " + directory);
+        }
+
+        IOException exception = null;
+        for (File file : files) {
+            try {
+                forceDelete(file);
+            } catch (IOException ioe) {
+                exception = ioe;
+            }
+        }
+
+        if (exception != null) {
+            throw exception;
+        }
+    }
+
+    private static void forceDelete(File file) throws IOException {
+        if (file.isDirectory()) {
+            deleteDirectory(file);
+        } else {
+            boolean filePresent = file.exists();
+            if (!file.delete()) {
+                if (!filePresent) {
+                    throw new FileNotFoundException("File does not exist: " + file);
+                }
+                String message =
+                        "Unable to delete file: " + file;
+                throw new IOException(message);
             }
         }
     }

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -1263,6 +1263,17 @@ public class FrontendUtils {
         deleteDirectory(nodeModules);
     }
 
+    /**
+     * Recursively delete given directory and contents.
+     * <p>
+     * Will not delete contents of symlink or junction directories, only the
+     * link file.
+     *
+     * @param directory
+     *            directory to delete
+     * @throws IOException
+     *             on failure to delete or read any one file
+     */
     public static void deleteDirectory(File directory) throws IOException {
         if (!directory.exists() || !directory.isDirectory()) {
             return;

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -1268,7 +1268,8 @@ public class FrontendUtils {
             return;
         }
 
-        if (!(Files.isSymbolicLink(directory.toPath()) || isJunction(directory.toPath()))) {
+        if (!(Files.isSymbolicLink(directory.toPath())
+                || isJunction(directory.toPath()))) {
             cleanDirectory(directory);
         }
 
@@ -1279,14 +1280,20 @@ public class FrontendUtils {
     }
 
     /**
-     * Check that directory is not a windows junction which is basically a symlink.
-     * @param directory directory path to check
+     * Check that directory is not a windows junction which is basically a
+     * symlink.
+     *
+     * @param directory
+     *            directory path to check
      * @return true if directory is a windows junction
-     * @throws IOException if an I/O error occurs
+     * @throws IOException
+     *             if an I/O error occurs
      */
     private static boolean isJunction(Path directory) throws IOException {
-        boolean isWindows = System.getProperty("os.name").toLowerCase().contains("windows");
-        BasicFileAttributes attrs = Files.readAttributes(directory, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
+        boolean isWindows = System.getProperty("os.name").toLowerCase()
+                .contains("windows");
+        BasicFileAttributes attrs = Files.readAttributes(directory,
+                BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
         return isWindows && attrs.isDirectory() && attrs.isOther();
     }
 
@@ -1302,7 +1309,7 @@ public class FrontendUtils {
         }
 
         File[] files = directory.listFiles();
-        if (files == null) {  // null if security restricted
+        if (files == null) { // null if security restricted
             throw new IOException("Failed to list contents of " + directory);
         }
 
@@ -1327,10 +1334,10 @@ public class FrontendUtils {
             boolean filePresent = file.exists();
             if (!file.delete()) {
                 if (!filePresent) {
-                    throw new FileNotFoundException("File does not exist: " + file);
+                    throw new FileNotFoundException(
+                            "File does not exist: " + file);
                 }
-                String message =
-                        "Unable to delete file: " + file;
+                String message = "Unable to delete file: " + file;
                 throw new IOException(message);
             }
         }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -421,7 +421,7 @@ public class FrontendUtilsTest {
         FrontendUtils.deleteNodeModules(new File(npmFolder, "node_modules"));
 
         Assert.assertTrue(
-                "File from linked folder was removed, even if only link folder should be touched.",
+                "Linked folder contents should not be removed.",
                 linkFolderFile.exists());
     }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -420,8 +420,7 @@ public class FrontendUtilsTest {
 
         FrontendUtils.deleteNodeModules(new File(npmFolder, "node_modules"));
 
-        Assert.assertTrue(
-                "Linked folder contents should not be removed.",
+        Assert.assertTrue("Linked folder contents should not be removed.",
                 linkFolderFile.exists());
     }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -373,7 +373,8 @@ public class FrontendUtilsTest {
     }
 
     @Test
-    public void symlinkByNpm_deleteDirectory_doesNotDeleteSymlinkFolderFiles() throws IOException, ExecutionFailedException {
+    public void symlinkByNpm_deleteDirectory_doesNotDeleteSymlinkFolderFiles()
+            throws IOException, ExecutionFailedException {
         File npmFolder = tmpDir.newFolder();
 
         File generatedPath = new File(npmFolder, "generated");
@@ -387,15 +388,15 @@ public class FrontendUtilsTest {
         final JsonObject packageJson = Json.createObject();
         packageJson.put(DEPENDENCIES, Json.createObject());
 
-        packageJson.getObject(DEPENDENCIES).put("@symbolic/link", "./" + symbolic.getName());
+        packageJson.getObject(DEPENDENCIES).put("@symbolic/link",
+                "./" + symbolic.getName());
 
         FileUtils.writeStringToFile(new File(npmFolder, PACKAGE_JSON),
                 packageJson.toJson(), StandardCharsets.UTF_8);
 
         ClassFinder finder = Mockito.mock(ClassFinder.class);
 
-        Logger logger = Mockito
-                .spy(LoggerFactory.getLogger(NodeUpdater.class));
+        Logger logger = Mockito.spy(LoggerFactory.getLogger(NodeUpdater.class));
 
         NodeUpdater nodeUpdater = new NodeUpdater(finder,
                 Mockito.mock(FrontendDependencies.class), npmFolder,
@@ -419,7 +420,9 @@ public class FrontendUtilsTest {
 
         FrontendUtils.deleteNodeModules(new File(npmFolder, "node_modules"));
 
-        Assert.assertTrue("File from linked folder was removed, even if only link folder should be touched.", linkFolderFile.exists());
+        Assert.assertTrue(
+                "File from linked folder was removed, even if only link folder should be touched.",
+                linkFolderFile.exists());
     }
 
     private ResourceProvider mockResourceProvider(VaadinService service) {

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -18,10 +18,12 @@ package com.vaadin.flow.server.frontend;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 
 import org.apache.commons.io.FileUtils;
@@ -31,21 +33,33 @@ import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 import org.mockito.Mockito;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import com.vaadin.experimental.FeatureFlags;
 import com.vaadin.flow.di.Lookup;
 import com.vaadin.flow.di.ResourceProvider;
 import com.vaadin.flow.function.DeploymentConfiguration;
+import com.vaadin.flow.server.ExecutionFailedException;
 import com.vaadin.flow.server.MockVaadinServletService;
 import com.vaadin.flow.server.ServiceException;
 import com.vaadin.flow.server.VaadinContext;
 import com.vaadin.flow.server.VaadinService;
 import com.vaadin.flow.server.VaadinServlet;
 import com.vaadin.flow.server.VaadinServletService;
+import com.vaadin.flow.server.frontend.installer.NodeInstaller;
+import com.vaadin.flow.server.frontend.scanner.ClassFinder;
+import com.vaadin.flow.server.frontend.scanner.FrontendDependencies;
 import com.vaadin.tests.util.MockDeploymentConfiguration;
 
+import elemental.json.Json;
+import elemental.json.JsonObject;
+import static com.vaadin.flow.server.Constants.PACKAGE_JSON;
 import static com.vaadin.flow.server.Constants.STATISTICS_JSON_DEFAULT;
+import static com.vaadin.flow.server.Constants.TARGET;
 import static com.vaadin.flow.server.Constants.VAADIN_SERVLET_RESOURCES;
 import static com.vaadin.flow.server.InitParameters.SERVLET_PARAMETER_STATISTICS_JSON;
+import static com.vaadin.flow.server.frontend.NodeUpdater.DEPENDENCIES;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
@@ -356,6 +370,56 @@ public class FrontendUtilsTest {
 
         Assert.assertFalse(nodeModules.exists());
         Assert.assertTrue(externalLicense.exists());
+    }
+
+    @Test
+    public void symlinkByNpm_deleteDirectory_doesNotDeleteSymlinkFolderFiles() throws IOException, ExecutionFailedException {
+        File npmFolder = tmpDir.newFolder();
+
+        File generatedPath = new File(npmFolder, "generated");
+        generatedPath.mkdir();
+
+        File symbolic = new File(npmFolder, "symbolic");
+        symbolic.mkdir();
+        File linkFolderFile = new File(symbolic, "symbol.txt");
+        linkFolderFile.createNewFile();
+
+        final JsonObject packageJson = Json.createObject();
+        packageJson.put(DEPENDENCIES, Json.createObject());
+
+        packageJson.getObject(DEPENDENCIES).put("@symbolic/link", "./" + symbolic.getName());
+
+        FileUtils.writeStringToFile(new File(npmFolder, PACKAGE_JSON),
+                packageJson.toJson(), StandardCharsets.UTF_8);
+
+        ClassFinder finder = Mockito.mock(ClassFinder.class);
+
+        Logger logger = Mockito
+                .spy(LoggerFactory.getLogger(NodeUpdater.class));
+
+        NodeUpdater nodeUpdater = new NodeUpdater(finder,
+                Mockito.mock(FrontendDependencies.class), npmFolder,
+                generatedPath, null, TARGET, Mockito.mock(FeatureFlags.class)) {
+
+            @Override
+            public void execute() {
+            }
+
+            @Override
+            Logger log() {
+                return logger;
+            }
+
+        };
+
+        new TaskRunNpmInstall(nodeUpdater, false, false,
+                FrontendTools.DEFAULT_NODE_VERSION,
+                URI.create(NodeInstaller.DEFAULT_NODEJS_DOWNLOAD_ROOT), false,
+                false, Collections.emptyList()).execute();
+
+        FrontendUtils.deleteNodeModules(new File(npmFolder, "node_modules"));
+
+        Assert.assertTrue("File from linked folder was removed, even if only link folder should be touched.", linkFolderFile.exists());
     }
 
     private ResourceProvider mockResourceProvider(VaadinService service) {

